app/src/main/java/com/github/onlynotesswent/model/folder/FolderRepositoryFirestore.kt

```diff
@@ -199,39 +350,120 @@ class FolderRepositoryFirestore(private val db: FirebaseFirestore) : FolderRepos
-                noteViewModel = noteViewModel,
-                onSuccess = {},
-                onFailure = {
-                  onFailure(it)
-                  Log.e(TAG, "Failed to delete folder contents: ${it.message}")
-                })
-            noteViewModel.deleteNotesFromFolder(subFolder.id)
-            deleteFolderById(
-                folderId = subFolder.id,
-                onSuccess = {},
-                onFailure = {
-                  onFailure(it)
-                  Log.e(TAG, "Failed to delete folderContents: ${it.message}")
-                })
+            CoroutineScope(Dispatchers.IO).launch {
+              deleteFolderContents(
+                  folder = subFolder,
+                  noteViewModel = noteViewModel,
+                  onSuccess = {},
+                  onFailure = {
+                    onFailure(it)
+                    Log.e(TAG, "Failed to delete folder contents: ${it.message}")
+                  },
+                  useCache = useCache)
+              noteViewModel.deleteNotesFromFolder(subFolder.id)
+              deleteFolderById(
+                  folderId = subFolder.id,
+                  onSuccess = {},
+                  onFailure = {
+                    onFailure(it)
+                    Log.e(TAG, "Failed to delete folderContents: ${it.message}")
+                  },
+                  useCache = useCache)
+            }
           }
           onSuccess()
         },
         onFailure = { e: Exception ->
-          onFailure(e)
           Log.e(TAG, "Failed to delete folder contents: ${e.message}")
-        })
+          onFailure(e)
+        },
+        useCache = useCache)
+  }
+
+  /**
+   * Synchronizes the Firestore database with the local cache based on `lastModified` field. Should
+   * be called only if the userId of the folder is the current user.
+   *
+   * @param firestoreFolder The folder from Firestore.
+   * @param cachedFolder The folder from the local cache.
+   * @return Latest version of the folder.
+   */
+  private suspend fun syncFolderFirestoreWithCache(
+      firestoreFolder: Folder,
+      cachedFolder: Folder?
+  ): Folder {
+    val updatedFolder =
+        if (cachedFolder == null || firestoreFolder.lastModified > cachedFolder.lastModified)
+            firestoreFolder // Firestore has the newest data
+        else cachedFolder // Local database has the newest data
+
+    // Update firestore and the cache with newest data
+    addFolder(updatedFolder, {}, {}, true)
+
+    return updatedFolder
+  }
+
+  /**
+   * Synchronizes the Firestore database with the local cache based on `lastModified` field. Should
+   * be called only if the userId of the folders is the current user.
+   *
+   * @param firestoreFolders The list of folders from Firestore.
+   * @param cachedFolders The list of folders from the local cache.
+   * @return Latest version of the list of folders.
+   */
+  private suspend fun syncFoldersFirestoreWithCache(
```

Comment:
Why not just cache the folder if necessary, instead of calling `addFolders` which caches it + updates firebase (not useful, since the cache is never going to be more updated than the firestore file as we have a write-through cache system: whenever you update a folder you update it in firestore)
This way we avoid a redundant call to the repository

Reply:
See above comment. Since they are related to the same topic, I will close this one and we can further discuss there if necessary